### DISCUSSION TITLE
Recommended State Object Type

### DISCUSSION THREAD

**Discussion by Besler** [Kitware/trame#216] [Q&A]
Is there a recommended object for managing the state variable schemas?

For instance, I typically use `pydantic.BaseModel` for most of my python code.
```python
from pydantic import BaseModel

class User(BaseModel):
    name: str
    location: str
    
user = User(name="Bryce", location="Canada")
print(user.json())
print(User.schema())
```

This way when I read or have to extend my code, I know what I'm looking at and what's expected.

Is there a recommended class for managing this?
```python
server.state.user = User(name="Bryce", location="Canada")
```

Potential features include:
- JSON serializable
- On JSON deserialization, parses to type
- Nesting
- Allows the addition of private methods on class

Ideally I like `pydantic` because of its integration with `sqlalchemy` and `FastAPI`. Is it worth me just hacking it together or would this be a worthwhile feature over all? I guess this is really a request to add types.

**Comment by jourdain** [comment]
I guess we could have it so you can drop any custom object (for you with type) assuming we provide an interface you need to implement (i.e. `get_state`, `set_state`) that will be handled transparently over the network/state-change and so on...

Would that work for you? If so would you be up for creating a PR with some guidance?

**Reply by Besler** [reply]
That sounds excellent - yes, can do.

**Comment by jourdain** [comment]
I'll try to layout the tasks for you with pointer on where some code should go later today so you can start working on it...
Thanks again for doing that!

**Comment by jourdain** [comment]
1. [ ] Create interface to implement (`CustomStateEntry:get_state/set_state`) [in here](https://github.com/Kitware/trame-server/blob/master/trame_server/state.py)
2. [ ] You won't be able to call [update](https://github.com/Kitware/trame-server/blob/master/trame_server/state.py#L170) directly. Instead you will need to go throw each entry and check if it is a `CustomStateEntry` and if so call `set_state`.
3. [ ] Update [def clean_value(value)](https://github.com/Kitware/trame-server/blob/master/trame_server/utils/__init__.py#L59) to use the `get_state` for `CustomStateEntry` value.
4. [ ] Provide a validation example showing that it work and what is needed. (trame/examples/validation/...)

The output/input of `get_state/set_state` should be serializable like a str or dict (json friendly).

I'm probably missing something, but that could be it... Let me know how it goes.
